pkg/sql/logictest/tests/fakedist-disk/fakedist-disk_test: TestLogic_datetime failed

pkg/sql/logictest/tests/fakedist-disk/fakedist-disk_test.TestLogic_datetime [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8375867?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8375867?buildTab=artifacts#/) on master @ [f27fa9d4c56e4654939e772d857027da00feb7d5](https://github.com/cockroachdb/cockroach/commits/f27fa9d4c56e4654939e772d857027da00feb7d5):


```
=== RUN   TestLogic_datetime
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f70ceff7ba074534c17c8b5bbdfe4516/logTestLogic_datetime3399248907
    test_log_scope.go:79: use -show-logs to present logs inline
[18:06:26] setting distsql_workmem='97583B';
=== RUN   TestLogic_datetime/regression_41663
=== RUN   TestLogic_datetime/regression_42244
=== RUN   TestLogic_datetime/tz_utc_normalization
=== RUN   TestLogic_datetime/infinity_time
=== RUN   TestLogic_datetime/interval_math
=== RUN   TestLogic_datetime/regression_35255
=== RUN   TestLogic_datetime/regression_36146
=== RUN   TestLogic_datetime/datestyle_order
[18:06:28] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11410/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/datetime with config fakedist-disk: 334 tests, 0 failures
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1a8 pc=0x2158288]

goroutine 31492 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x400663ed48?, {0x59e2008, 0x4002053470})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x68
panic({0x3ee0ba0, 0x8103e60})
	GOROOT/src/runtime/panic.go:884 +0x20c
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cancel(0x400411b880?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:527 +0x18
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupFlows.func5({0x4003ba7000?, 0x464cf0?})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:625 +0x1f0
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x128
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x358
=== RUN   TestLogic_datetime/regression_36557
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_datetime.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23570

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -369,6 +369,10 @@ func (f *vectorizedFlow) MemUsage() int64 {
 
 // Cleanup is part of the flowinfra.Flow interface.
 func (f *vectorizedFlow) Cleanup(ctx context.Context) {
+	startCleanup, endCleanup := f.FlowBase.GetOnCleanupFns()
+	startCleanup()
+	defer endCleanup()
+
 	// This cleans up all the memory and disk monitoring of the vectorized flow
 	// as well as closes all the closers.
 	f.creator.cleanup(ctx)

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -216,7 +216,7 @@ func (ds *ServerImpl) setupFlow(
 ) (retCtx context.Context, _ flowinfra.Flow, _ execopnode.OpChains, retErr error) {
 	var sp *tracing.Span          // will be Finish()ed by Flow.Cleanup()
 	var monitor *mon.BytesMonitor // will be closed in Flow.Cleanup()
-	var onFlowCleanup func()
+	var onFlowCleanupEnd func()   // will be called at the very end of Flow.Cleanup()
 	// Make sure that we clean up all resources (which in the happy case are
 	// cleaned up in Flow.Cleanup()) if an error is encountered.
 	defer func() {
@@ -224,8 +224,8 @@ func (ds *ServerImpl) setupFlow(
 			if monitor != nil {
 				monitor.Stop(ctx)
 			}
-			if onFlowCleanup != nil {
-				onFlowCleanup()
+			if onFlowCleanupEnd != nil {
+				onFlowCleanupEnd()
 			} else {
 				reserved.Close(ctx)
 			}
@@ -307,7 +307,7 @@ func (ds *ServerImpl) setupFlow(
 		// the whole evalContext, but that isn't free, so we choose to restore
 		// the original state in order to avoid performance regressions.
 		origTxn := evalCtx.Txn
-		onFlowCleanup = func() {
+		onFlowCleanupEnd = func() {
 			evalCtx.Txn = origTxn
 			reserved.Close(ctx)
 		}
@@ -322,7 +322,7 @@ func (ds *ServerImpl) setupFlow(
 			evalCtx.Txn = leafTxn
 		}
 	} else {
-		onFlowCleanup = func() {
+		onFlowCleanupEnd = func() {
 			reserved.Close(ctx)
 		}
 		if localState.IsLocal {
@@ -388,7 +388,7 @@ func (ds *ServerImpl) setupFlow(
 	isVectorized := req.EvalContext.SessionData.VectorizeMode != sessiondatapb.VectorizeOff
 	f := newFlow(
 		flowCtx, sp, ds.flowRegistry, rowSyncFlowConsumer, batchSyncFlowConsumer,
-		localState.LocalProcs, isVectorized, onFlowCleanup, req.StatementSQL,
+		localState.LocalProcs, isVectorized, onFlowCleanupEnd, req.StatementSQL,
 	)
 	opt := flowinfra.FuseNormally
 	if !localState.MustUseLeafTxn() {
@@ -521,10 +521,10 @@ func newFlow(
 	batchSyncFlowConsumer execinfra.BatchReceiver,
 	localProcessors []execinfra.LocalProcessor,
 	isVectorized bool,
-	onFlowCleanup func(),
+	onFlowCleanupEnd func(),
 	statementSQL string,
 ) flowinfra.Flow {
-	base := flowinfra.NewFlowBase(flowCtx, sp, flowReg, rowSyncFlowConsumer, batchSyncFlowConsumer, localProcessors, onFlowCleanup, statementSQL)
+	base := flowinfra.NewFlowBase(flowCtx, sp, flowReg, rowSyncFlowConsumer, batchSyncFlowConsumer, localProcessors, onFlowCleanupEnd, statementSQL)
 	if isVectorized {
 		return colflow.NewVectorizedFlow(base)
 	}

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -43,7 +43,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
-	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
 	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -100,23 +99,25 @@ type runnerResult struct {
 // run executes the request. An error, if encountered, is both sent on the
 // result channel and returned.
 func (req runnerRequest) run() error {
-	defer physicalplan.ReleaseFlowSpec(&req.flowReq.Flow)
 	res := runnerResult{nodeID: req.sqlInstanceID}
+	defer func() {
+		req.resultChan <- res
+		physicalplan.ReleaseFlowSpec(&req.flowReq.Flow)
+	}()
 
 	conn, err := req.podNodeDialer.Dial(req.ctx, roachpb.NodeID(req.sqlInstanceID), rpc.DefaultClass)
 	if err != nil {
 		res.err = err
-	} else {
-		client := execinfrapb.NewDistSQLClient(conn)
-		// TODO(radu): do we want a timeout here?
-		resp, err := client.SetupFlow(req.ctx, req.flowReq)
-		if err != nil {
-			res.err = err
-		} else {
-			res.err = resp.Error.ErrorDetail(req.ctx)
-		}
+		return err
 	}
-	req.resultChan <- res
+	client := execinfrapb.NewDistSQLClient(conn)
+	// TODO(radu): do we want a timeout here?
+	resp, err := client.SetupFlow(req.ctx, req.flowReq)
+	if err != nil {
+		res.err = err
+	} else {
+		res.err = resp.Error.ErrorDetail(req.ctx)
+	}
 	return res.err
 }
 
@@ -464,12 +465,13 @@ func (dsp *DistSQLPlanner) setupFlows(
 
 	// Start all the remote flows.
 	//
-	// numAsyncRequests tracks the number of the SetupFlow RPCs that were
-	// delegated to the DistSQL runner goroutines.
-	var numAsyncRequests int
-	// numSerialRequests tracks the number of the SetupFlow RPCs that were
-	// issued by the current goroutine on its own.
-	var numSerialRequests int
+	// usedWorker indicates whether we used at least one DistSQL worker
+	// goroutine to issue the SetupFlow RPC.
+	var usedWorker bool
+	// numIssuedRequests tracks the number of the SetupFlow RPCs that were
+	// issued (either by the current goroutine directly or delegated to the
+	// DistSQL workers).
+	var numIssuedRequests int
 	if sp := tracing.SpanFromContext(origCtx); sp != nil && !sp.IsNoop() {
 		setupReq.TraceInfo = sp.Meta().ToProto()
 	}
@@ -515,7 +517,7 @@ func (dsp *DistSQLPlanner) setupFlows(
 			//
 			// Note that even in case of an error in runnerRequest.run we still
 			// send on the result channel.
-			for i := 0; i < numAsyncRequests+numSerialRequests; i++ {
+			for i := 0; i < numIssuedRequests; i++ {
 				<-resultChan
 			}
 			// At this point, we know that all concurrent requests (if there
@@ -541,11 +543,11 @@ func (dsp *DistSQLPlanner) setupFlows(
 
 		// Send out a request to the workers; if no worker is available, run
 		// directly.
+		numIssuedRequests++
 		select {
 		case dsp.runnerCoordinator.runnerChan <- runReq:
-			numAsyncRequests++
+			usedWorker = true
 		default:
-			numSerialRequests++
 			// Use the context of the local flow since we're executing this
 			// SetupFlow RPC synchronously.
 			runReq.ctx = ctx
@@ -554,16 +556,8 @@ func (dsp *DistSQLPlanner) setupFlows(
 			}
 		}
 	}
-	if buildutil.CrdbTestBuild {
-		if numAsyncRequests+numSerialRequests != len(flows)-1 {
-			return ctx, flow, errors.AssertionFailedf(
-				"expected %d requests, found only %d async and %d serial",
-				len(flows)-1, numAsyncRequests, numSerialRequests,
-			)
-		}
-	}
 
-	if numAsyncRequests == 0 {
+	if !usedWorker {
 		// We executed all SetupFlow RPCs in the current goroutine, and all RPCs
 		// succeeded.
 		return ctx, flow, nil

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -580,7 +580,7 @@ func (dsp *DistSQLPlanner) setupFlows(
 		syncutil.Mutex
 		called bool
 	}{}
-	flow.AddOnCleanup(func() {
+	flow.AddOnCleanupStart(func() {
 		cleanupCalledMu.Lock()
 		defer cleanupCalledMu.Unlock()
 		cleanupCalledMu.called = true
@@ -599,24 +599,29 @@ func (dsp *DistSQLPlanner) setupFlows(
 		for i := 0; i < len(flows)-1; i++ {
 			res := <-resultChan
 			if res.err != nil && !seenError {
-				seenError = true
 				// The setup of at least one remote flow failed.
-				cleanupCalledMu.Lock()
-				skipCancel := cleanupCalledMu.called
-				cleanupCalledMu.Unlock()
-				if skipCancel {
-					continue
-				}
-				// First, we update the DistSQL receiver with the error to be
-				// returned to the client eventually.
-				//
-				// In order to not protect DistSQLReceiver.status with a mutex,
-				// we do not update the status here and, instead, rely on the
-				// DistSQLReceiver detecting the error the next time an object
-				// is pushed into it.
-				recv.setErrorWithoutStatusUpdate(res.err, true /* willDeferStatusUpdate */)
-				// Now explicitly cancel the local flow.
-				flow.Cancel()
+				seenError = true
+				func() {
+					cleanupCalledMu.Lock()
+					// Flow.Cancel cannot be called after or concurrently with
+					// Flow.Cleanup.
+					defer cleanupCalledMu.Unlock()
+					if cleanupCalledMu.called {
+						// Cleanup of the local flow has already been performed,
+						// so there is nothing to do.
+						return
+					}
+					// First, we update the DistSQL receiver with the error to
+					// be returned to the client eventually.
+					//
+					// In order to not protect DistSQLReceiver.status with a
+					// mutex, we do not update the status here and, instead,
+					// rely on the DistSQLReceiver detecting the error the next
+					// time an object is pushed into it.
+					recv.setErrorWithoutStatusUpdate(res.err, true /* willDeferStatusUpdate */)
+					// Now explicitly cancel the local flow.
+					flow.Cancel()
+				}()
 			}
 		}
 	})

--- a/pkg/sql/flowinfra/flow.go
+++ b/pkg/sql/flowinfra/flow.go
@@ -140,10 +140,14 @@ type Flow interface {
 	// Cleanup.
 	Cancel()
 
-	// AddOnCleanup adds a callback to be executed at the very end of Cleanup.
-	// Callbacks are put on the stack meaning that AddOnCleanup is called
-	// multiple times, then the "later" callbacks are executed first.
-	AddOnCleanup(fn func())
+	// AddOnCleanupStart adds a callback to be executed at the very beginning of
+	// Cleanup.
+	AddOnCleanupStart(fn func())
+
+	// GetOnCleanupFns returns a couple of functions that should be called at
+	// the very beginning and the very end of Cleanup, respectively. Both will
+	// be non-nil.
+	GetOnCleanupFns() (startCleanup, endCleanup func())
 
 	// Cleanup must be called whenever the flow is done (meaning it either
 	// completes gracefully after all processors and mailboxes exited or an
@@ -200,7 +204,10 @@ type FlowBase struct {
 	//  - outboxes
 	waitGroup sync.WaitGroup
 
-	onFlowCleanup func()
+	// onCleanupStart and onCleanupEnd will be called in the very beginning and
+	// the very end of Cleanup(), respectively.
+	onCleanupStart func()
+	onCleanupEnd   func()
 
 	statementSQL string
 
@@ -276,7 +283,7 @@ func NewFlowBase(
 	rowSyncFlowConsumer execinfra.RowReceiver,
 	batchSyncFlowConsumer execinfra.BatchReceiver,
 	localProcessors []execinfra.LocalProcessor,
-	onFlowCleanup func(),
+	onFlowCleanupEnd func(),
 	statementSQL string,
 ) *FlowBase {
 	// We are either in a single tenant cluster, or a SQL node in a multi-tenant
@@ -300,7 +307,7 @@ func NewFlowBase(
 		batchSyncFlowConsumer: batchSyncFlowConsumer,
 		localProcessors:       localProcessors,
 		admissionInfo:         admissionInfo,
-		onFlowCleanup:         onFlowCleanup,
+		onCleanupEnd:          onFlowCleanupEnd,
 		status:                flowNotStarted,
 		statementSQL:          statementSQL,
 	}
@@ -527,17 +534,31 @@ func (f *FlowBase) Cancel() {
 	f.ctxCancel()
 }
 
-// AddOnCleanup is part of the Flow interface.
-func (f *FlowBase) AddOnCleanup(fn func()) {
-	if f.onFlowCleanup != nil {
-		oldOnFlowCleanup := f.onFlowCleanup
-		f.onFlowCleanup = func() {
+// AddOnCleanupStart is part of the Flow interface.
+func (f *FlowBase) AddOnCleanupStart(fn func()) {
+	if f.onCleanupStart != nil {
+		oldOnCleanupStart := f.onCleanupStart
+		f.onCleanupStart = func() {
 			fn()
-			oldOnFlowCleanup()
+			oldOnCleanupStart()
 		}
 	} else {
-		f.onFlowCleanup = fn
+		f.onCleanupStart = fn
 	}
+}
+
+var noopFn = func() {}
+
+// GetOnCleanupFns is part of the Flow interface.
+func (f *FlowBase) GetOnCleanupFns() (startCleanup, endCleanup func()) {
+	onCleanupStart, onCleanupEnd := f.onCleanupStart, f.onCleanupEnd
+	if onCleanupStart == nil {
+		onCleanupStart = noopFn
+	}
+	if onCleanupEnd == nil {
+		onCleanupEnd = noopFn
+	}
+	return onCleanupStart, onCleanupEnd
 }
 
 // Cleanup is part of the Flow interface.
@@ -594,9 +615,6 @@ func (f *FlowBase) Cleanup(ctx context.Context) {
 	}
 	f.status = flowFinished
 	f.ctxCancel()
-	if f.onFlowCleanup != nil {
-		f.onFlowCleanup()
-	}
 }
 
 // cancel cancels all unconnected streams of this flow. This function is called

--- a/pkg/sql/rowflow/row_based_flow.go
+++ b/pkg/sql/rowflow/row_based_flow.go
@@ -439,6 +439,9 @@ func (f *rowBasedFlow) Release() {
 
 // Cleanup is part of the flowinfra.Flow interface.
 func (f *rowBasedFlow) Cleanup(ctx context.Context) {
+	startCleanup, endCleanup := f.FlowBase.GetOnCleanupFns()
+	startCleanup()
+	defer endCleanup()
 	f.FlowBase.Cleanup(ctx)
 	f.Release()
 }

